### ISSUE TITLE
EDDI 3.3.7- b1 seems to bypass 'say' script -wait until complete' -setting

### ISSUE THREAD

**Issue by FireSammy** [EDCD/EDDI#1245]
## EDDI version in which issue found
EDDI 3.3.7- b1 used with VA 1.7.3

## Steps to reproduce
Really hard to say exactly. Some speech responder scripts finish ok as usual, and some seem to get faded (VA fade stopped audio -selected) before finished. Many of my commands/command strings have both (wait until finished) set, and some not have it selected.

## Expected
EDDI to wait for the speech script to finish when waiting has been set. If not set, then not stop the audio to be faded, but queue it just like it used to do. This has also important timing value for command strings to fire in correct sequence as planned, instead of everything having a timed pause in between. I'm sure you know what i mean.

## Observed
Currently, in some cases it does not seem to have any difference do i have a speech script waiting on or off... VA fading speech and double speech for same script occurs sometimes -not often but on rare occasions. Seems to me that the trouble to wait for the scripts makes these problems happening. Might also be a VA problem too. Made a separate issues post in VA forums.

I hope i don't get kicked from both lol. I don't have time to do deeper tests for investigations. Sorry

**Comment by Tkael** [comment]
Which scripts?

**Comment by Tkael** [comment]
Which scripts? There are a couple (for example, `Hull damage`, that are intended to automatically interrupt when newer data becomes available).

**Comment by FireSammy** [comment]
Priority settings?? I'm talking about VA text scripts for EDDI to speak (say).
EDDI's own event (speech) scripts are not the issue here (i think?) lol

**Comment by Tkael** [comment]
This is unlikely to be anything on VoiceAttack's side.

**Comment by FireSammy** [comment]
ok

**Comment by FireSammy** [comment]
Erhm... quoting "VoiceAttack say and speak commands will use the optional priority parameter"
I don't see any optional parameter for priority here (check pic) and i don't bloody well want to see one either... unless it is set to highest by default. EDDI API does seem to have this feature and everything is set to 3 by default. Hard to say if this is messing with my command "say' scripts then? 

Otherwise, I have quite a list of commands to redo if a separate priority thingy will be added here and will virtually make a complete mess on my profile as this issue posted above is already doing...

![Priority setting](https://user-images.githubusercontent.com/21980416/54475083-f8e7a100-47f5-11e9-9692-ed932c4c81c1.JPG)

**Comment by FireSammy** [comment]
Now that i'm thinking about it, it seems that me saying it is "bypassing" my 'say' scripts - is too mildly said. It seems to completely override them and stopping them. That is why i get the voice faded away because VA considers EDDI talking script stopped. Overriding will of course then also affect the waiting before continue -setting

I would hope this new priority scale would make the EDDI speech responder to prioritise what will be said first etc. and not to completely override others with no priority setting... Well thats just my point of view. I welcome all new setting features, but i hope this "overriding" thing was not intended? If i understood this problem correct that is... If it's a bug, well then it's just an issue for me and others who are using both types of speech scripts for spoken stuff

o7

**Comment by Tkael** [comment]
See here.
https://github.com/EDCD/EDDI/blob/develop/docs/VoiceAttack-Integration.md#say

Use of priorities 2-4 is advised for most applications. Priority 5 speech is interruptible by any higher priority speech. Priority 1 will interrupt any lower priority speech.  Speech from VoiceAttack defaults to 3 but can be changed as documented above.

**Comment by FireSammy** [comment]
Much better at priority set to 5 -BUT

Still overriding occurs and indeed waiting for the VA 'say' command script to finnish before continuing will get overwritten as well. This will have some unexpected effect on timing command strings -not even mention the missed voice scripts being overwritten.

Is this really working as intended?

**Comment by Tkael** [comment]
Priority 3 (the default) is the best choice for maintaining traditional speech behavior.

**Comment by FireSammy** [comment]
Again just making sure any misunderstanding does not happen. I really think this priority setting is a great thing. Many will find it good and handy. I will too, when i get into it.

Adding a priority bypass option will make it 100% = Everyone wins. My humble wish in other words.

With scale from 1-5 FORCED priority effect is overriding stuff and will make some of us spend hours and hours to go trough say -script commands adding pauses into them... I repeat: It overrides the VA function -'wait for the plugin function to finnish before continuing' at least when it comes to 'say' scripts.

o7

**Comment by FireSammy** [comment]
> Priority 3 (the default) is the best choice for maintaining traditional speech behavior, as that matches the VoiceAttack default.

Negative. For crying out loud... I'm sorry i have to repeat myself here. I had the EDDI API default values and it overwrote every VA function -'wait for the plugin function to finnish before continuing' at least when it comes to 'say' scripts in event triggered command strings. The value 5 helped a lot, but it still ovverrides things like described above.

**Comment by Tkael** [comment]
`wait for the plugin function to finish before continuing` describes whether VoiceAttack waits for control to be returned, not whether EDDI waits. 

We had an issue where the `((EDDI Jumped))` was triggering late because of other bottlenecked speech (which was very undesirable). Our goal is (and must remain) to offload onto a separate thread and return control to VoiceAttack as soon as possible so that we don't slow / bottleneck other VoiceAttack commands.

**Comment by Tkael** [comment]
We can look at whether we can respect the VoiceAttack checkbox value, but in the meantime EDDI sets a value in VoiceAttack to indicate when it is speaking. That value can be used to make a VoiceAttack command wait for speech to complete. 
https://github.com/EDCD/EDDI/wiki/VoiceAttack-Integration#miscellaneous-variables

**Comment by Tkael** [comment]
After reviewing the VoiceAttack docs, it doesn't look like we can support this as an optional parameter - the content of that checkbox isn't communicated via the plugin.

**Comment by Tkael** [comment]
How we should respond if EDDI is already speaking or had speech queued when this speech is invoked? Should we block until all speech is complete? Until just the next speech is complete?

**Comment by FireSammy** [comment]
Huh oki, i humbly thank you for your attention to this. 
This is good {BOOL:EDDI speaking}: true if EDDI is currently speaking -i have to figure out how to use it then. But it's good to have this available. Adding 'priority' instead of 'say' -to script i assume it will set the priority to default 3 -very good oki.

Problems are still clear because of this current thing: When two,- or more ((EDDI event))'s are triggered in short sequence after another -they will override the earlier one's speech responders voice text even when/if these events have the same priority number selected, like i have them now at 5.

I think having two events with the same priority number should not override the other one, no matter if they are set to1 or 5. Currently it does at least when having it at 5. And it did when i had it on 3.

So i don't have any easy/fast control over this now. All i can do currently is simply add timed pauses into each place i can expect overlapping to occur. I already did last night for some 4 hours and got at least the most important EDDI events back under reasonable control.

Still, there are situations where multiple events are triggered shortly after another (such as approaching to a station and requesting docking), and there i simply have to delete all EDDI API speech responders which i know for sure that will overlap and disrupt any previous event speech, and make it somewhat similar voice script into the event command to get it at least somewhat under timing control.

Quite a lot of work in other words, which is quite inconvenient as these command strings have been working very well for over 2 years now. This priority setting is surely good in a combat situation when things can get queued really fast when many events are being triggered as well and all of them are usually a high priority. I can easily understand that. Other things going on outside combat zone is usually quite the opposite and is important to have command strings reliably proceed with timed sequence(s)

I understand bypassing this priority setting the way i suggested is not an option. Is it possible to make sure the events triggered with same priority number would not override the previous events speech? This is my biggest question here.

**Comment by Tkael** [comment]
New speech should never interrupt active speech of the same priority. 
- Can you give a specific example of where you are seeing this? 
- Are you invoking VoiceAttack speech or using EDDI scripts in the cases where you are observing this?

**Comment by Tkael** [comment]
Hrm. I just got back to my dev pc after traveling for the past several days. I'm having trouble replicating your observations using the following VA commands:
```
Set Text [Script] to 'Testing speech'
Execute external plugin, 'EDDI 3.3.7-b2' and wait for return
Set Text [Script] to 'Testing speech 2'
Execute external plugin, 'EDDI 3.3.7-b2'
```
Result is "Testing speech. Testing speech 2."

Do you have a VA command sequence where "wait for return" is being ignored so that I can better replicate this?

**Comment by Tkael** [comment]
I went ahead and submitted a small tweak so that EDDI won't return control to VA until the speech invoked by VA is completed.

**Comment by MadHoek** [comment]
Hi there, 
I got kind of the same problem using the most recent version of EDDI (3.3.7-b1).
I use "say scripts" and they work fine for me no interruptions or overlapping after said scripts I run actions. The speech finishes in a queue but the actions run independently from them as fast as somehow possible pauses or wait till the speech finishes seem to be ignored (for a lot of them I created a VA function e.g. "no voice"-commands -> no "when I press keys" or "when I say" ticked). This wasn't the case using EDDi 3.3.6.

Example:
say: blah blah -> wait to finish
run action 1 (open a program for example)
say: blah blah blah -> wait to finish
run action 2 (open another program for example)
say: blah blah blah blah -> wait to finish
run action 3 (open some other program for example)

Result:
say: blah blah -> ongoing
run action 1 (open a program for example)
run action 2 (open another program for example)
run action 3 (open some other program for example)
say: blah blah -> finishes
say: blah blah blah -> finishes
say: blah blah blah blah -> finishes

I hope you get what I mean and I can help a bit with this.

Cheers
